src/agent/src/policy.rs

```diff
@@ -0,0 +1,218 @@
+        // Delete the old rules.
+        self.opa_client
+            .delete(&self.policy_path)
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        // Put the new rules.
+        self.opa_client
+            .put(&self.policy_path)
+            .body(policy.to_string())
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        // Check if requests causing policy errors should actually be allowed.
+        // That is an insecure configuration but is useful for allowing insecure
+        // pods to start, then connect to them and inspect Guest logs for the
+        // root cause of a failure.
+        self.allow_failures = self
+            .post_query("AllowRequestsFailingPolicy", EMPTY_JSON_INPUT)
+            .await?;
+        Ok(())
+    }
+
+    // Post query to OPA.
+    async fn post_query(&mut self, ep: &str, post_input: &str) -> Result<bool> {
+        debug!(sl!(), "policy check: {}", ep);
+
+        let uri = self.query_path.clone() + ep;
+        let response = self
+            .opa_client
+            .post(uri)
+            .body(post_input.to_string())
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        if response.status() != http::StatusCode::OK {
+            return Err(anyhow!(
+                "policy: post_query: POST response status {}",
+                response.status()
+            ));
+        }
+
+        let http_response = response.text().await.unwrap();
+        let opa_response: serde_json::Result<AllowResponse> = serde_json::from_str(&http_response);
+
+        match opa_response {
+            Ok(resp) => {
+                if !resp.result {
+                    if self.allow_failures {
+                        warn!(
+                            sl!(),
+                            "policy: post_query: response <{}>. Ignoring error!", http_response
+                        );
+                        return Ok(true);
+                    } else {
+                        error!(sl!(), "policy: post_query: response <{}>", http_response);
+                    }
+                }
+                Ok(resp.result)
+            }
+            Err(_) => {
+                warn!(
+                    sl!(),
+                    "policy: post_query: {} not found in policy. Returning false.", ep,
+                );
+                Ok(false)
+            }
+        }
+    }
+
+    #[cfg(feature = "agent-policy-debug")]
+    async fn log_opa_input(&mut self, ep: &str, opa_input: &str) {
+        if let Some(log_file) = &mut self.log_file {
+            match ep {
+                "StatsContainerRequest" | "ReadStreamRequest" | "SetPolicyRequest" => {}
+                _ => {
+                    let log_entry = "# ".to_string() + ep + "\n\n" + opa_input + "\n\n";
+                    log_file.write_all(log_entry.as_bytes()).await.unwrap();
+                    log_file.flush().await.unwrap();
```

Comment:
Please try to avoid using `unwrap()`. Change the function definition to `...-> Result<()>` is a choice. In this way all the errors occuring inside can be raised to upper caller.

Reply:
I replaced the unwrap calls, but I added an abort call - please see https://github.com/kata-containers/kata-containers/pull/7574/commits/5f1bbe4a79d3d6d07ec25e0253c23b3f6dc758b0